File: leetcode/0630-course-schedule-iii/solution.py
import heapq
from typing import List


class Solution:
    def scheduleCourse(self, courses: List[List[int]]) -> int:
        # take courses in deadline order; whenever the running total blows past
        # a deadline, kick out the single most time-consuming course taken so
        # far (not necessarily this one) since that frees up the most room
        courses.sort(key=lambda c: c[1])

        max_heap = []  # negated durations of courses currently "taken"
        total_time = 0

        for duration, deadline in courses:
            heapq.heappush(max_heap, -duration)
            total_time += duration

            if total_time > deadline:
                total_time += heapq.heappop(max_heap)  # heappop gives -largest, so this subtracts it

        return len(max_heap)

File: leetcode/0630-course-schedule-iii/solution.cpp
#include <vector>
#include <algorithm>
#include <queue>
using namespace std;

class Solution {
public:
    int scheduleCourse(vector<vector<int>>& courses) {
        sort(courses.begin(), courses.end(), [](const vector<int>& a, const vector<int>& b) {
            return a[1] < b[1];
        });

        priority_queue<int> maxHeap; // max-heap of durations taken so far
        long long totalTime = 0;

        for (auto& course : courses) {
            int duration = course[0];
            int deadline = course[1];

            maxHeap.push(duration);
            totalTime += duration;

            if (totalTime > deadline) {
                totalTime -= maxHeap.top();
                maxHeap.pop();
            }
        }

        return (int)maxHeap.size();
    }
};
